[BC-breaking] Fix bugs in torch::tensor constructor

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28524 [WIP] Support C++ tensor advanced indexing
* **#28523 [BC-breaking] Fix bugs in torch::tensor constructor**

New features:
1. Previously, `torch::tensor({true, false, true})` throws `"tensor_cpu" not implemented for 'Bool'`. After this PR, it produces the correct bool tensor, matching the Python API behavior.
2. Tensors with zero-size dimensions are now supported, e.g. `torch::tensor({{}, {}})` produces a tensor with sizes `{2, 0}`, matching the Python API behavior.

BC-breaking bug fixes and changes:
1. Previously, `torch::tensor({{1}, {2}})` produces a tensor of sizes `{2}`. After this PR, it produces a tensor of sizes `{2, 1}`, matching the Python API behavior.
2. Fixed semantics of `torch::tensor(1.1)`: it now returns a 0-dim tensor instead of a 1-dim tensor, matching the Python API behavior.
3. After this PR, passing `std::initializer_list` (NOT braced-init-list) to `torch::tensor` doesn't work anymore, and the user should pass the equivalent braced-init-list to `torch::tensor` instead. For example, `torch::tensor(std::initializer_list<double>({1.1, 1.2}))` doesn't work anymore, and the user should do `torch::tensor({1.1, 1.2})`. (The reason for this change is that `TensorDataContainer` cannot have two constructor that takes `std::initializer_list`, otherwise a value such as `{1.1, 1.2}` would be ambiguous - it can take the `std::initializer_list<TensorDataContainer>` constructor, or it can take the `std::initializer_list<double>` constructor.)
4. Previously, when passed a non-dtype `TensorOptions` to the `torch::tensor` constructor, it always produces a tensor of dtype `float`. After this PR, it produces tensor of different dtypes based on the dtype of the braced-init-list, matching the behavior of the no-options case.
```cpp
// Previously:
torch::tensor({1, 2, 3}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> float
torch::tensor({{1, 2, 3}}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> float
torch::tensor({1., 2., 3.}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> float
torch::tensor({{1., 2., 3.}}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> float

// Now:
torch::tensor({1, 2, 3}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> int
torch::tensor({{1, 2, 3}}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> int
torch::tensor({1., 2., 3.}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> double
torch::tensor({{1., 2., 3.}}, torch::TensorOptions(/*non-dtype-options*/)).dtype() -> double

// As comparison, currently:
torch::tensor({1, 2, 3}).dtype() -> int
torch::tensor({{1, 2, 3}}).dtype() -> int
torch::tensor({1., 2., 3.}).dtype() -> double
torch::tensor({{1., 2., 3.}}).dtype() -> double
```

Notes:
1. From now on, the behavior of `at::tensor(scalar_value)` (which produces a 1-dim tensor) would be different from `torch::tensor(scalar_value)` (which produces a 0-dim tensor). I will fix the behavior of `at::tensor(scalar_value)` in a follow-up PR.
2. From now on, the behavior of `at::tensor({1, 2, 3}, torch::TensorOptions(/*non-dtype-options*/))` (which produces a `float` tensor) would be different from `torch::tensor({1, 2, 3}, torch::TensorOptions(/*non-dtype-options*/))` (which produces a an `int` tensor). I will fix this behavior of `at::tensor` constructor in a follow-up PR.

Context for the changes in this PR:

The motivation comes from fixing the "`torch::tensor({{1}, {2}})` gives tensor of wrong sizes" bug - in order to fix it, I have to move the handling of `at::ArrayRef` and `std::vector` into `InitListTensor` (see below on why we need to do this) and renamed `InitListTensor` to `TensorDataContainer`. After such changes, support for bool values comes out of the box without extra effort, and support for tensors with zero-size dimensions only requires adding a default constructor for `TensorDataContainer`, so I added those two in this PR.  

For the semantic change of `torch::tensor(1.1)`, it's actually more effort to preserve the original wrong behavior (i.e. we need to check the sizes of the tensor converted from `TensorDataContainer` and reshape any scalar tensor to a 1-D tensor). I think preserving the original wrong behavior doesn't give us much value, and since the above changes naturally fix the problem, we should just start using the right behavior instead.

For the "constructor with non-dtype options behavior" fix, the code looks simpler and easier to reason about with the fix, so I included it in this PR.

--------

Why we need to move the handling of `at::ArrayRef` and `std::vector` into `TensorDataContainer`:

`torch::tensor({{1}, {2}})` can match this function overload:
`torch::tensor(at::ArrayRef<int> values)`, because `{1}` and `{2}` can be treated as
a list-initialization of an `int` value. However, this will produce a Tensor with sizes `{2}`,
but we actually want a Tensor with sizes `{2, 1}`. In order to avoid matching this function overload,
we removed the function overload and moved the ability to convert `at::ArrayRef<T>`
(and similarly `std::vector<T>`) into `TensorDataContainer`, and since for braced-init-list the
`TensorDataContainer(std::initializer_list<TensorDataContainer>)` constructor is always preferred over all other constructors, it will take the `std::initializer_list` path, and all is good.

Differential Revision: [D18234625](https://our.internmc.facebook.com/intern/diff/D18234625)

Commit:
Fix bugs in torch::tensor constructor

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

yf225 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -22,80 +22,170 @@ using at::DimnameList;
+  const c10::Scalar& scalar() const {
+    TORCH_CHECK(
+      is_scalar(),
+      "Can only call `scalar()` on a TensorDataContainer that has `is_scalar() == true`");
```
TODO: I need to add tests for these error messages

yf225 replied:
Fixed

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -22,80 +22,170 @@ using at::DimnameList;
-        scalar_type_(at::k##S),        \
-        type_(InitListTensorType::Scalar) {}
+
+#define NESTED_INIT_LIST_MAX_DEPTH 10
```
This is better as `constexpr int64_t kNestedInitListMaxDepth = 10`

yf225 replied:
Thanks! I refactored the PR and got rid of the need for this const variable.

zou3519 commented:
Could you walk through how in the original code, `torch::tensor({{1}, {2}})` produced a tensor of size `{2}`?

zou3519 commented:
> Could you walk through how in the original code, `torch::tensor({{1}, {2}})` produced a tensor of size `{2}`?

This is what I think was what was going on in the previous code (please let me know if this is right or not):

{1} becomes a InitListTensor of scalar 1
{2} becomes an InitListTensor of scalar 2
{ InitListTensor(1), InitListTensor(2) } becomes an InitListTensor with sizes [2]

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -22,80 +22,170 @@ using at::DimnameList;
+
+#define NESTED_INIT_LIST_MAX_DEPTH 10
+
+enum class TensorDataContainerType { Scalar, InitList, Tensor };
```
The code to support the `torch::tensor `constructor is getting really long. Would it be better to toss it into its own file?

yf225 replied:
I moved it to `torch/csrc/api/include/torch/detail/TensorDataContainer.h`.

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -104,117 +194,164 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, TENSOR)
-  inline at::Tensor tensor(T value) {                                      \
-    return torch::tensor(at::ArrayRef<T>(value));                          \
+  std::vector<int64_t> sizes_;
+  c10::ScalarType scalar_type_;
```
Have you considered having TensorDataContainer hold `std::variant<Scalar, Tensor, initializer_list<TensorDataContainer<D+1>>>` in this implementation ?  Would it make the implementation look nicer, worse, or roughly the same?

zou3519 replied:
https://en.cppreference.com/w/cpp/utility/variant says "A variant is not permitted to hold references, arrays, or the type void." and I guess initializer_list is a type of array?

yf225 replied:
I am a little worried that accessing the member value of `c10::variant` would be slower than accessing it as a direct class member. Because we want to prevent perf regression on the existing `torch::tensor(at::ArrayRef<T>)` and `torch::tensor(std::vector<T>)` as much as possible, I feel that it might be better to store them as direct class members.

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -22,80 +22,170 @@ using at::DimnameList;
+// to always prefer the scalar constructor `TensorDataContainer(int)` instead of the `std::initializer_list`
+// constructor, thus producing a tensor with the wrong sizes. If we templatize `TensorDataContainer` over
+// the # of tensor dimensions, such problem doesn't happen.
+template <size_t D>
```
What is D? The depth? Could we name it something more descriptive?

yf225 replied:
I refactored the PR and got rid of the need for this template variable :D

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -22,80 +22,170 @@ using at::DimnameList;
+//
+// which calls:
+//
+// `TensorDataContainer<1>({{1}, {2}})`
```
Skimming through the code, it looks like it only matters if `D = 1` or if `D > 1` because we define some constructors on `D = 1`. Is this true?

yf225 replied:
I refactored the PR and got rid of the need for this const variable. :D

zou3519 reviewed (commented):
Still reading, have some initial questions

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"


New features:
1. Previously, `torch::tensor({true, false, true})` throws `"tensor_cpu" not implemented for 'Bool'`. After this PR, it produces the correct bool tensor, matching the Python API behavior.
2. Tensors with zero-size dimensions are now supported, e.g. `torch::tensor({{}, {}})` produces a tensor with sizes `{2, 0}`, matching the Python API behavior.

Bug fixes:
1. Previously, `torch::tensor({{1}, {2}})` produces a tensor of sizes `{2}`. After this PR, it produces a tensor of sizes `{2, 1}`, matching the Python API behavior.
2. Fixed semantics of `torch::tensor(1.1)`: it now returns a 0-dim tensor instead of a 1-dim tensor, matching the Python API behavior.

BC-breaking changes:
1. `torch::tensor(1.1)` now returns a 0-dim tensor instead of a 1-dim tensor, matching the Python API behavior.

Notes:
1. From now on, the behavior of `at::tensor(scalar_value)` (which produces a 1-dim tensor) would be different from `torch::tensor(scalar_value)` (which produces a 0-dim tensor). I will fix the behavior of `at::tensor(scalar_value)` in the next PR.

The motivation comes from fixing the "`torch::tensor({{1}, {2}})` gives tensor of wrong sizes" bug - in order to fix it, I have to create the templated `TensorDataContainer` class and move the handling of `at::ArrayRef` and `std::vector` into it. After such changes, support for bool values comes out of the box without extra effort, and support for tensors with zero-size dimensions only requires adding a default constructor for `TensorDataContainer`, so I added those two in this PR.  

For the semantic change of `torch::tensor(1.1)`, it's actually more effort to preserve the original wrong behavior (i.e. we need to explicitly instantiate `TensorDataContainer<1>` and let its scalar constructor create a 1-D tensor). I think preserving the original wrong behavior doesn't give us much value, and since the above changes naturally fix the problem, we should just start using the right behavior instead.

[ghstack-poisoned]

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

yf225 commented:
> > Could you walk through how in the original code, `torch::tensor({{1}, {2}})` produced a tensor of size `{2}`?
> 
> This is what I think was what was going on in the previous code (please let me know if this is right or not):
> 
> {1} becomes a InitListTensor of scalar 1
> {2} becomes an InitListTensor of scalar 2
> { InitListTensor(1), InitListTensor(2) } becomes an InitListTensor with sizes [2]

I made a mistake in my original assessment, and here is what actually happened instead:

`torch::tensor({{1}, {2}})` matches a previously existing function overload: `torch::tensor(at::ArrayRef<int> values)`, because the compiler looks at  `{1}` and `{2}` as
list-initializations of `int` values, and produce a Tensor with sizes `{2}`.

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

yf225 commented on test/cpp/api/autograd.cpp:
```diff
@@ -518,8 +518,8 @@ TEST(CustomAutogradTest, ReentrantPriority) {
-  auto a = MyFunction::apply(torch::tensor(6, torch::requires_grad()));
-  auto b = Reenter::apply(torch::tensor(9, torch::requires_grad()));
+  auto a = MyFunction::apply(torch::tensor({6}, torch::dtype(torch::kFloat).requires_grad(true)));
+  auto b = Reenter::apply(torch::tensor({9}, torch::dtype(torch::kFloat).requires_grad(true)));
```
I changed these tensors from 0-D to 1-D and explicitly passed the dtype, so that the original behavior of the tests are preserved.

yf225 commented on test/cpp/api/dataloader.cpp:
```diff
@@ -616,8 +616,8 @@ struct UnCopyableDataset : public datasets::Dataset<UnCopyableDataset> {
-    return {torch::tensor(static_cast<int64_t>(index)),
-            torch::tensor(static_cast<int64_t>(index))};
+    return {torch::tensor({static_cast<int64_t>(index)}),
+            torch::tensor({static_cast<int64_t>(index)})};
```
I changed these tensors from 0-D to 1-D, so that the original behavior of the tests are preserved.

yf225 commented on test/cpp/api/functional.cpp:
```diff
@@ -88,7 +88,7 @@ TEST_F(FunctionalTest, CosineSimilarity) {
 
 TEST_F(FunctionalTest, SoftMarginLossDefaultOptions) {
-  auto input = torch::tensor({2., 4., 1., 3.}, torch::requires_grad());
+  auto input = torch::tensor({2., 4., 1., 3.}, torch::dtype(torch::kFloat).requires_grad(true));
```
I explicitly passed the dtype to all `torch::tensor` calls that use options, so that the resulting tensor's dtype is always what we expect it to be.

yf225 commented on test/cpp/api/parallel.cpp:
```diff
@@ -212,7 +212,7 @@ TEST_F(ParallelTest, DataParallelUsesAllAvailableCUDADevices_CUDA) {
     void reset() override {}
     torch::Tensor forward(torch::Tensor input) {
-      return torch::tensor(input.device().index());
+      return torch::tensor({input.device().index()});
```
I changed these tensors from 0-D to 1-D, so that the original behavior of the tests are preserved.

zou3519 commented on test/cpp/api/tensor.cpp:
```diff
@@ -211,6 +260,15 @@ TEST(TensorTest, ContainsCorrectValuesForManyValuesVariable) {
   ASSERT_TRUE(tensor.is_variable());
   ASSERT_EQ(tensor.numel(), 3);
+  ASSERT_EQ(tensor.sizes(), std::vector<int64_t>({3}));
+  ASSERT_EQ(tensor.dtype(), at::kDouble);
```
To match the python API, I'd expect both of these to have dtype float32. Is there a reason why we aren't doing this?

yf225 replied:
Yes I haven't got around to add `torch::set_default_dtype` yet. I will do it as a follow-up PR.

zou3519 replied:
How hard is it to add torch::set_default_dtype? It looks like it might be more efficient to add it and then merge this PR (so we don't have to change dtypes manually to kFloat in this change and then reverse this change in the future)

yf225 replied:
I found that we already have `torch::set_default_dtype` implemented in https://github.com/pytorch/pytorch/blob/d8c368bd62455f2c9d0bf036640a87281db0face/c10/core/DefaultDtype.h#L10
but to change the default dtype to float32 for the `torch::tensor` constructor would be another BC-breaking change, which I think could happen independent of this PR. (When we change the default dtype in that PR, we will only need to change the expected default dtype for test cases such as `torch::tensor({1.5, 2.25, 3.125})` in `tensor.cpp`, and we don't need to change the functional tests / module tests because we explicitly specify their input's dtype (which is good because it doesn't have dependency on what the default dtype is).)

zou3519 replied:
Got it, that sounds reasonable. Could we add a comment somewhere (I'm not sure where) about the inconsistency (torch::tensor by default give a double tensor) that links an issue that says we are going to work on changing torch::tensor to give a float tensor by default?

zou3519 replied:
Maybe in `torch::tensor`

yf225 replied:
Thanks! I added it above `inline at::Tensor tensor(detail::TensorDataContainer tensor_data_container, const at::TensorOptions& options = {})` in this PR

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+// but we actually want a Tensor with sizes `{2, 1}`. In order to avoid matching this function overload,
+// we removed the function overload and moved the ability to convert `at::ArrayRef<T>` (and similarly `std::vector<T>`)
+// into `TensorDataContainer`, and since for braced-init-list the `TensorDataContainer(std::initializer_list<TensorDataContainer>)`
+// constructor is always preferred over all other constructors, it will take the `std::initializer_list` path
```
nit: It's kinda hard to read lines on github when they're over 80-100 lines long (this one is 132), but don't worry about changing it

yf225 replied:
Fixed

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+    }
+  }
+ private:
+  void fill_tensor(at::Tensor tensor) const {
```
Better to pass by mutable reference so that (1) the semantics are clearer and (2) we avoid a refcount bump (those are expensive)

yf225 replied:
I found that I wasn't able to change this to mutable reference, because in `fill_tensor` we have this line `elem.fill_tensor(tensor[index])`, and `tensor[index]` returns a `Tensor` instead of a `Tensor&`.
```
../torch/csrc/api/include/torch/detail/TensorDataContainer.h:273:39: error: invalid initialization of non-const reference of type ‘at::Tensor&’ from an rvalue of type ‘at::Tensor’
         elem.fill_tensor(tensor[index]);
                                       ^
```

zou3519 replied:
Alternatively, I think the following might make the semantics clearer:
- use (Tensor& Tensor)
```
Tensor slice = tensor[index];
elem.fill_tensor(slice);
```

yf225 replied:
Thanks a lot for the suggestions! I tried them out and the second approach works well for all `fill_tensor` calls in this file :D

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+    if (is_scalar()) {
+      at::NoGradGuard guard;
+      tensor.fill_(scalar_);
+    } else if (is_init_list()) {
```
We do no error checking in this function -- is the assumption that `tensor` is always the correct size? If so, could we add some asserts?

yf225 replied:
Thanks for the suggestion! I added asserts to this function

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+#pragma once
```
Sometimes it is nicer to split declarations and definitions into .h and .cpp files. i.e., definitions go into .h files and declarations go into .cpp files. There are probably a lot of considerations about this that I am missing, but here are some pros and cons:

- (pro) compilation speeds: when one is making changes to TensorDataContainer.cpp, they don't need to recompile every file that that TensorDataContainer.h is included in
- (pro) modularity: Staring at the .h file gives a good sense of the interface 
- (con) Need to add a new file to the build system. Probably not the worse thing in the world

yf225 replied:
Yes I think it's a good idea to split it into .h and .cpp files, although when I tried to do this I ran into a lot of "undefined reference" errors that I am still trying to find the reason for. I will attempt the split again when I implement the multidim support for the `at::tensor` constructor.

yf225 replied:
Thanks a lot for the tip! I am gonna try that again in the next patch :/

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -21,198 +22,15 @@ using at::DimnameList;
-
-inline at::Tensor tensor(std::initializer_list<detail::InitListTensor> init_list) {
-  return torch::tensor(init_list, at::dtype(init_list.begin()->scalar_type()));
+inline at::Tensor tensor(detail::TensorDataContainer init_list_tensor, const at::TensorOptions& options = {}) {
```
If I'm understanding this correctly, this is the only overload for `torch::tensor`. In that case, maybe it's better to call the variable `tensor_data_container` since it doesn't have to be an initialization list anymore?

yf225 replied:
Fixed

zou3519 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -21,198 +22,15 @@ using at::DimnameList;
-  return torch::tensor(init_list, at::dtype(init_list.begin()->scalar_type()));
+inline at::Tensor tensor(detail::TensorDataContainer init_list_tensor, const at::TensorOptions& options = {}) {
+  return autograd::make_variable(
+    init_list_tensor.convert_to_tensor(options.has_dtype() ? options : options.dtype(init_list_tensor.scalar_type())),
```
nit: I feel like we could hide this detail (deciding to use `options.dtype` or `init_list_tensor.scalar_type()` inside convert_to_tensor

yf225 replied:
Done :D

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+        ", but got element of scalar type: ",
+        elem.scalar_type());
+    }
+    sizes_.clear();
```
Is the clear() necessary?

yf225 replied:
No it's not - removed

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+  }
+
+#define TENSOR(T, S) \
+  TensorDataContainer(const at::ArrayRef<T>& values) { \
```
I think ArrayRef is meant to be passed by value

yf225 replied:
Fixed

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+  TensorDataContainer(const at::ArrayRef<T>& values) { \
+    type_ = TensorDataContainerType::Tensor; \
+    sizes_ = {(int64_t)values.size()}; \
+    scalar_type_ = at::k##S; \
```
Can these not go into the constructor?
```
TensorDataContainer(const at::ArrayRef<T>& values) : type_(...), sizes_(...) {
```
or does the macro hackery prevent that?

yf225 replied:
Thanks for the catch! I fixed it :D

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+  // NOTE: `torch::tensor(std::vector<bool>)` is not supported for now, because
+  // ArrayRef<bool> cannot be constructed from a std::vector<bool> bitfield.
+#define TENSOR(T, S) \
+  TensorDataContainer(std::vector<T> values) : TensorDataContainer(at::ArrayRef<T>(values)) {}
```
Should be `const std::vector<T>&`, we don't want to copy the vector here

yf225 replied:
Fixed

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+
+  void pretty_print_recursive(std::ostream& stream) const {
+    if (is_scalar()) {
+      // NOTE: There is no `operator<<` overload for `at::kBFloat16` type,
```
Should we add a `operator<<` overload for `at::kBFloat16`? cc @izdeby

izdeby replied:
Yes, we should. i will file an issue

yf225 replied:
Added

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,272 @@
+      }
+    } else if (is_init_list()) {
+      stream << "{";
+      for (const TensorDataContainer* it = init_list_.begin(); it != init_list_.end(); it++) {
```
nit: you could `const auto` here but this is fine

zou3519 replied:
oh, I see. "auto" or "const auto*" should work but it's fine to keep it as `const TensorDataContainer*`

zou3519 reviewed (commented):
I need to go think a bit more about the correctness of this but here are some small comments

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,303 @@
+//
+// `torch::tensor({{1}, {2}})`
+//
+// Here is the code path that it goes through:
```
It's a little weird to say "here is the codepath that it goes through" because there is only one torch::tensor overload. Maybe say "this is the torch::tensor declaration"?

yf225 replied:
Fixed

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,303 @@
+//
+// Now it matches `TensorDataContainer(int value)`, which stores `1` as a scalar value. All is good.
+//
+// Note that `torch::tensor({{1}, {2}})` can also match another previously existing function overload:
```
We don't really need this paragraph (which is a retrospective on why the old approach didn't work). Something like this is good to have in the commit message / pull request body but devs trying to understand how TensorDataConstructor works only need to know that it is used for calling `torch::tensor` with ArrayRef, vectors, and braced initialization lists.

yf225 replied:
Removed the paragraph :D

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,303 @@
+        // and we need to convert it to `float` type using `operator float()` function
+        // defined in c10/util/BFloat16.h.
+        if (scalar_type_ == at::ScalarType::BFloat16) {
+          stream << static_cast<float>(tensor_[i].item<c10::BFloat16>());
```
It's not-so-nice to have the note in two places. Maybe write a helper function 
`FIXME_dump_bfloat16(const Tensor& scalar, ostream& blah)`?

yf225 replied:
I added helper function `std::ostream& operator<<(std::ostream& stream, c10::BFloat16 value)` and move the note to be above the helper function

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,303 @@
+              at::kHalf,
+              scalar_type_,
+              "TensorDataContainer_pretty_print_tensor_item", [&] {
+            stream << tensor_[i].item<scalar_t>();
```
Does `stream << tensor_[i].item()` work without dispatching?

yf225 replied:
`item()` requires a template type `T` so I think we will need to pass a `scalar_t`. I also tried `stream << tensor_[i]` but that prints
```
1.1
[ CPUFloatType{} ]
```
instead of `1.1`, which is probably not nice for the pretty print.

zou3519 commented on test/cpp/api/tensor.cpp:
```diff
@@ -191,10 +191,58 @@ TEST(TensorTest, ContainsCorrectValuesForManyValues) {
+  ASSERT_TRUE(almost_equal(tensor[0], true));
+}
+
 TEST(TensorTest, ContainsCorrectValuesForManyValuesVariable) {
```
"Many values" doesn't mean much to me, could we call this "TensorCtorSingleDim" or something?

yf225 replied:
Renamed

zou3519 commented on test/cpp/api/tensor.cpp:
```diff
@@ -219,20 +277,84 @@ TEST(TensorTest, ContainsCorrectValuesForManyValuesVariable) {
 }
 
 TEST(TensorTest, MultidimTensorCtor) {
+  {
```
Can we split this test into smaller tests?

i.e., we could have a "zero-sized dim test" which are all of the {} / {{},{}} ones

Also, it might be good to assert the dtype of torch::tensor({}) (I'm not sure if that test exists already)

yf225 replied:
Added `TensorCtorZeroSizedDim` and also fixed the code to let tensor with zero-sized dim to be of `float` dtype by default, matching the Python API behavior.

zou3519 commented on test/cpp/api/tensor.cpp:
```diff
@@ -304,27 +478,81 @@ TEST(TensorTest, MultidimTensorCtor_CUDA) {
+      c10::str(torch::detail::TensorDataContainer({{{{{{{{{{1, 2}}}}}}}}}})),
+      "{{{{{{{{{{1, 2}}}}}}}}}}");
+  }
+}
```
I don't think you have a test for pretty-printing a TensorDataContainer that holds a tensor

yf225 replied:
Thanks for the catch! Fixed.

yf225 replied:
Added the tests (which caught a printing bug and I just fixed it :D)

zou3519 reviewed (commented):
The approach looks correct to me. I had some minor comments on cosmetics and testing

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

yf225 commented:
@zou3519 Thanks so much for the reviews! I address all of them and please feel free to look at it again when you have time :D

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,293 @@
+  }
+
+#define TENSOR(T, S) \
+  TensorDataContainer(at::ArrayRef<T> values) : \
```
Just checking, we don't (and we didn't) support `torch::tensor(tensor)` before, right?

yf225 replied:
Right we didn't support it, and we currently don't support it either.

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,293 @@
+      scalar_type_(at::k##S), \
+      type_(TensorDataContainerType::Tensor) { \
+    at::AutoNonVariableTypeMode non_var_type_mode(true);  \
+    tensor_ = at::tensor(values, at::TensorOptions().device(at::kCPU).is_variable(false));       \
```
What are the differences in semantics between at::tensor and torch::tensor? Does `at::tensor` only take in ArrayRefs?

yf225 replied:
I believe `at::tensor` takes in scalar value / ArrayRef / std::vector / braced-init-list at the moment:
https://github.com/pytorch/pytorch/blob/d6f1e49c4a3d853edf18a53e827d648541fac44a/aten/src/ATen/templates/NativeFunctions.h#L30-L49

zou3519 replied:
ah, and at::tensor(1.) returns a 1D tensor instead of a scalar tensor. Do you think it would be worth cleaning up at::tensor to look more like torch::tensor at some point?

yf225 replied:
Yes I just filed two issues on this topic:
https://github.com/pytorch/pytorch/issues/28929
https://github.com/pytorch/pytorch/issues/28930
It would be a good idea to unify the behavior of `at::tensor` and `torch::tensor`.

zou3519 commented on torch/csrc/api/include/torch/detail/TensorDataContainer.h:
```diff
@@ -0,0 +1,293 @@
+        tensor.sizes()[0],
+        " in its first dimension");
+      size_t index = 0;
+      for (const auto& elem : init_list_) {
```
Does std::initializer_list support indexing? if so it looks like it might be nicer as a for loop: `for (size_t index = 0; index < init_list_.size(); ++index)`

yf225 replied:
I thought about the same too, but the author of C++ said it doesn't support subscripting :( https://stackoverflow.com/a/17823064

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

yf225 commented on test/cpp/api/tensor.cpp:
```diff
@@ -191,10 +191,58 @@ TEST(TensorTest, ContainsCorrectValuesForManyValues) {
 }
 
-TEST(TensorTest, ContainsCorrectValuesForManyValuesVariable) {
+TEST(TensorTest, ContainsCorrectValueForSingleValueVariable) {
```
I renamed these tests and also refactored some others to make the test structure cleaner :D

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

Commit:
Update on "[BC-breaking] Fix bugs in torch::tensor constructor"

Fix bugs in torch::tensor constructor

gh-metadata: pytorch pytorch 28523 gh/yf225/16/head

kostmo commented:
## CircleCI build failures summary
As of commit ef6898b:
* 0/2 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=ef6898bb8a0759c2c27ffb3708d0cfce190852fe).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
